Classes/common/touchdb/ChangeTracker/TDURLConnectionChangeTracker.m

```diff
@@ -0,0 +1,298 @@
+            CDTLogError(CDTREPLICATION_LOG_CONTEXT,
+                       @"%@: HTTP auth failed; sent Authorization: %@  ;  got WWW-Authenticate: %@", [self class],
+                       authorization, authResponse);
+            errorInfo = $dict({ @"HTTPAuthorization", authorization },
+                              { @"HTTPAuthenticateHeader", authResponse });
+        }
+        
+        //retryOrError will only retry if the error seems to be a transient error.
+        //otherwise, retryOrError will set the error and stop.
+        [self retryOrError:TDStatusToNSErrorWithInfo(status, self.changesFeedURL, errorInfo)];
+    }
+    
+}
+
+-(void) connection:(NSURLConnection *)connection didReceiveData:(NSData *)data
+{
+    CDTLogVerbose(CDTREPLICATION_LOG_CONTEXT, @"%@: didReceiveData: %ld bytes",
+                  [self class], (unsigned long)[data length]);
+    
+    [self.inputBuffer appendData:data];
+}
+
+-(void) connectionDidFinishLoading:(NSURLConnection *)connection
+{
+    //parse the input buffer into JSON (or NSArray of changes?)
+    CDTLogVerbose(CDTREPLICATION_LOG_CONTEXT, @"%@: didFinishLoading, %u bytes", self,
+               (unsigned)self.inputBuffer.length);
+    
+    BOOL restart = NO;
+    NSString* errorMessage = nil;
+    NSInteger numChanges = [self receivedPollResponse:self.inputBuffer errorMessage:&errorMessage];
+    
+    if (numChanges < 0) {
+        // unparseable response. See if it gets special handling:
+        if ([self receivedDataBeginsCorrectly:self.inputBuffer]) {
+            
+            // The response at least starts out as what we'd expect, so it looks like the connection
+            // was closed unexpectedly before the full response was sent.
+            NSTimeInterval elapsed = [self.startTime timeIntervalSinceNow] * -1000.0;
+            CDTLogError(CDTREPLICATION_LOG_CONTEXT, @"%@: connection closed after %.1f sec",
+                        self, elapsed);
+            
+            [self retryOrError:[NSError errorWithDomain:NSURLErrorDomain
+                                                   code:NSURLErrorNetworkConnectionLost
+                                               userInfo:nil]];
+            
+            return;
+        }
+        
+        // Otherwise report an upstream unparseable-response error
+        [self setUpstreamError:errorMessage];
+    }
+    else {
+        // Poll again if there was no error, and it looks like we
+        // ran out of changes due to a _limit rather than because we hit the end.
+        restart = numChanges == (NSInteger)_limit;
+    }
+    
+    [self clearConnection];
+    
+    if (restart)
+        [self start];  // Next poll...
+    else
+        [self stopped];
+    
+}
+
+-(void) connection:(NSURLConnection *)connection didFailWithError:(NSError *)error
+{
+    [self retryOrError:error];
+}
+
+- (BOOL)receivedDataBeginsCorrectly:(NSData*)body
+{
+    // Convert to string:
+    NSString* bodyStr = [body my_UTF8ToString];
```

Comment:
I'm a bit worried about loading the entire response into a string -- if we're in an error state it could be larger than we expect. We know we should have UTF8 text here, we can just read the buffer itself:

```objc
	NSString* inputString = @"    {\"results\":";
	NSData* inputData = [inputString dataUsingEncoding:NSUTF8StringEncoding];
	NSUInteger inputLength = [inputData length];
	
	NSData *prefixData = [@"{\"results\":" dataUsingEncoding:NSUTF8StringEncoding];
	int prefixLength = [prefixData length];
	
	for (NSUInteger index = 0; index < inputLength; index++)
	{
		char currentChar;
		NSRange currentCharRange = NSMakeRange(index, 1);
		[inputData getBytes:&currentChar range:currentCharRange];
	
		// If it's the opening {, check for valid start JSON
	   	if (currentChar == '{') {
			NSRange r = NSMakeRange(index, prefixLength);
			char buf[prefixLength];
			
			if (inputLength >= (index + prefixLength)) {
				[inputData getBytes:buf range:r];
				BOOL match = (memcmp(buf, prefixData.bytes, prefixLength) == 0);
				if (match) {
					NSLog(@"found result start JSON");
				}
			}
			break;  // once we've seen a {, break always as can't succeed if we've not
					// already.
		}
	}
```